File: solutions/scala/wordy/1/src/main/scala/Wordy.scala
object Wordy {
  val functionByWord: Map[String, (Int, Int) => Int] = Map[String, (Int, Int) => Int](
    "plus" -> add,
    "minus" -> subtract,
    "multiplied" -> multiply,
    "divided" -> divide
  )

  def add(a: Int, b: Int): Int = a + b

  def subtract(a: Int, b: Int): Int = a - b

  def multiply(a: Int, b: Int): Int = a * b

  def divide(a: Int, b: Int): Int = a / b

  def answer(question: String): Option[Int] = {
    this.answer(question.split(" ").toList)
  }

  private def answer(words: List[String]): Option[Int] = {
    words match {
      case "What" :: "is" :: tail =>
        this.evaluate(tail)

      case _ =>
        None
    }
  }

  private def evaluate(words: List[String]): Option[Int] = {
    this.extract_first_words(words) match {
      case Some((firstNumber, operation, secondNumber, tail)) =>
        this.evaluate(firstNumber, operation, secondNumber, tail)

      case _ =>
        None
    }
  }

  private def extract_first_words(words: List[String]): Option[(Int, String, Int, List[String])] = {
    words match {
      case firstNumber :: operation :: secondNumber :: tail => {
        if (secondNumber == "by") {
          val secondNumber2 :: tail2 = tail
          Some((to_integer(firstNumber), operation, to_integer(secondNumber2), tail2))
        }
        else {
          Some((to_integer(firstNumber), operation, to_integer(secondNumber), tail))
        }
      }
      case _ =>
        None
    }
  }

  private def evaluate(
                        firstNumber: Int,
                        operation: String,
                        secondNumber: Int,
                        tail: List[String]
                      ): Option[Int] = {
    val result: Int = functionByWord(operation)(firstNumber, secondNumber)

    if (tail.nonEmpty) {
      this.answer(List("What", "is", result.toString) ::: tail)
    }
    else {
      Some(result)
    }
  }

  private def to_integer(value: String): Int = {
    value.replace("?", "")
      .toInt
  }
}

File: solutions/elixir/wordy/1/lib/wordy.ex
defmodule Wordy do
  @function_by_word %{
    "plus" => &__MODULE__.add/2,
    "minus" => &__MODULE__.subtract/2,
    "multiplied" => &__MODULE__.multiply/2,
    "divided" => &__MODULE__.divide/2
  }

  def add(a, b),
    do: a + b

  def subtract(a, b),
    do: a - b

  def multiply(a, b),
    do: a * b

  def divide(a, b),
    do: div(a, b)

  @doc """
  Calculate the math problem in the sentence.
  """
  @spec answer(String.t()) :: integer
  def answer(question)
      when is_binary(question),
      do:
        question
        |> String.split()
        |> answer()

  def answer(["What", "is" | [first_number, operation, "by", second_number | more]]),
    do: answer(first_number, operation, second_number, more)

  def answer(["What", "is" | [first_number, operation, second_number | more]]),
    do: answer(first_number, operation, second_number, more)

  defp answer(first_number, operation, second_number, more) do
    {first_value, _} = Integer.parse(first_number)
    {second_value, _} = Integer.parse(second_number)

    result = Map.fetch!(@function_by_word, operation).(first_value, second_value)

    if length(more) > 0,
      do: answer(["What", "is", to_string(result) | more]),
      else: result
  end

  def answer(_),
    do: raise(ArgumentError)
end
